File: wallet/src/main/java/com/worldventures/wallet/service/nxt/DetokenizeMultipleRecordsCommand.java
package com.worldventures.wallet.service.nxt;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.wallet.domain.entity.record.Record;
import com.worldventures.wallet.service.nxt.model.MultiRequestElement;
import com.worldventures.wallet.service.nxt.model.MultiResponseBody;
import com.worldventures.wallet.service.nxt.util.DetokenizedRecord;
import com.worldventures.wallet.service.nxt.util.NxtBankCardHelper;
import com.worldventures.wallet.service.nxt.util.NxtRecord;

import java.util.List;

import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class DetokenizeMultipleRecordsCommand extends BaseMultipleRecordsCommand {

   public DetokenizeMultipleRecordsCommand(List<? extends Record> records, boolean skipTokenizationErrors) {
      super(records, skipTokenizationErrors, false);
   }

   @Override
   List<MultiRequestElement> prepareMultiRequestElements(Record record) {
      return NxtBankCardHelper.getDataForDetokenization(record);
   }

   @Override
   List<NxtRecord> createResponseBody(List<Record> records, List<MultiResponseBody> nxtResponses) {
      return Queryable.from(records)
            .map(record -> DetokenizedRecord.from(record, nxtResponses))
            .cast(NxtRecord.class)
            .toList();
   }

}

File: wallet/src/main/java/com/worldventures/wallet/service/nxt/DetokenizeMultipleRecordsCommand.kt
package com.worldventures.wallet.service.nxt

import com.worldventures.wallet.domain.entity.record.Record
import com.worldventures.wallet.service.nxt.model.MultiRequestElement
import com.worldventures.wallet.service.nxt.model.MultiResponseBody
import com.worldventures.wallet.service.nxt.util.DetokenizedRecord
import com.worldventures.wallet.service.nxt.util.NxtBankCardHelper
import com.worldventures.wallet.service.nxt.util.NxtRecord
import io.techery.janet.command.annotations.CommandAction

@CommandAction
class DetokenizeMultipleRecordsCommand(records: List<Record>, skipTokenizationErrors: Boolean) : BaseMultipleRecordsCommand(records, skipTokenizationErrors, false) {

   internal override fun prepareMultiRequestElements(record: Record): List<MultiRequestElement> {
      return NxtBankCardHelper.getDataForDetokenization(record)
   }

   internal override fun createResponseBody(records: List<Record>, nxtResponses: List<MultiResponseBody>): List<NxtRecord> {
      return records
            .map { record -> DetokenizedRecord.from(record, nxtResponses) }
   }

}